### PR TITLE
Refactors isAllowedToLand code, adds checks for open tasks, allows co…

### PR DESCRIPTION
**Pull request by lukebatchelor**
…nfiguring checks in config, adds back in webhooks

Is good. Trust me

### CODE REVIEW

File: src/Client.js [hunk 1]
```diff
@@ -5,22 +5,79 @@ import type {
...
+import Logger from './Logger';
 
 export default class Client {
   hostAdaptor: Host;
   ciAdaptor: CI;
   persona: Persona;
+  serverSettings: Settings;
 
-  constructor(host: Host, ci: CI, persona: Persona) {
+  constructor(host: Host, ci: CI, persona: Persona, settings: Settings) {
```
Comment (lukebatchelor) [inline src/Client.js]: I don't love how all these things get passed around atm, but I'm happy for now

File: src/Client.js [hunk 2]
```diff
@@ -5,22 +5,79 @@ import type {
...
     this.hostAdaptor = host;
     this.ciAdaptor = ci;
     this.persona = persona;
+    this.serverSettings = settings;
   }
 
   async isAllowedToLand(pullRequestId: string) {
-    return await this.hostAdaptor.isAllowedToLand(pullRequestId);
```
Comment (lukebatchelor) [inline src/Client.js]: This was one of the main things.
Moving the "isAllowedToLand" logic out of the HostAdaptor.

This is much better

File: src/Client.js [hunk 3]
```diff
@@ -5,22 +5,79 @@ import type {
...
+      isAllowed = isAllowed && isApproved;
+    }
+    if (this.serverSettings.requireClosedTasks) {
+      isAllowed = isAllowed && allTasksClosed;
+    }
+    if (this.serverSettings.requireGreenBuild) {
+      isAllowed = isAllowed && isGreen;
+    }
+
+    Logger.info(
```
Comment (lukebatchelor) [inline src/Client.js]: I hate what prettier does to this. -.-

File: src/ci/BitbucketPipelinesAdapter.js [hunk 4]
```diff
@@ -33,20 +33,29 @@ const BitbucketPipelinesAdapter: CIAdapter = (config: Config) => {
     processStatusWebhook(body: JSONValue): StatusEvent | null {
       if (
         !body ||
-        !body.commit_status ||
-        !body.commit_status.state ||
-        !body.commit_status.url ||
-        typeof body.commit_status.url !== 'string'
+        !body.data ||
+        !body.data.commit_status ||
```
Comment (lukebatchelor) [inline src/ci/BitbucketPipelinesAdapter.js]: This is super strange.

Part of the mystery around the strange build_status' was solved, but it actually doesnt explain how this was ever working.

File: src/hosts/BitbucketAdapter.js [hunk 5]
```diff
@@ -129,21 +100,43 @@ const BitbucketAdapter = (config: Config) => {
...
+    async getPullRequest(pullRequestId: string): Promise<PullRequest> {
       const endpoint = `${apiBaseUrl}/pullrequests/${pullRequestId}`;
       const resp = await axios.get(endpoint, axiosGetConfig);
-      return resp.data;
+      const data = resp.data;
+      const approvals = data.participants
+        .filter(participant => participant.approved)
+        .map(participant => participant.user.username);
+
+      return {
```
Comment (lukebatchelor) [inline src/hosts/BitbucketAdapter.js]: Now we don't expose implementation details about BB api's, we decide our own shape

File: src/static/bitbucket/atlassian-connect.json [hunk 6]
```diff
@@ -29,6 +29,10 @@
           "permission": "write"
         }
       }]
+    }],
+    "webhooks": [{
```
Comment (lukebatchelor) [inline src/static/bitbucket/atlassian-connect.json]: Added these back.

Cant remember why they were removed. 🤷‍♂️

File: src/types.js [hunk 7]
```diff
@@ -85,10 +100,18 @@ export type CIConfig = {
   REPO_SLUG: string
 };
 
+export type Settings = {
+  requireApproval: boolean,
+  requireClosedTasks: boolean,
+  requireGreenBuild: boolean,
+  usersAllowedToApprove: Array<string>
```
Comment (lukebatchelor) [inline src/types.js]: I'll likely move the `usersAllowedToApprove` here at some point. This was just getting ahead of myself

**Comment by lukebatchelor** [comment]
Im happy to merge this now, or when you have a look @d4rkr00t

File: src/Client.js [hunk 8]
```diff
@@ -5,22 +5,79 @@ import type {
...
+      {
+        pullRequestId,
+        isOpen,
+        isApproved,
+        isGreen,
+        allTasksClosed
+      },
+      'isAllowedToLand()'
+    );
+    console.log('this.serverSettings', this.serverSettings);
```
Comment (d4rkr00t) [inline src/Client.js]: console.log :)

File: src/Client.js [hunk 9]
```diff
@@ -5,22 +5,79 @@ import type {
...
+    const pullRequest: PullRequest = await this.hostAdaptor.getPullRequest(
+      pullRequestId
+    );
+    const buildStatuses = await this.hostAdaptor.getPullRequestBuildStatuses(
+      pullRequestId
+    );
+
+    const isOpen = pullRequest.state === 'OPEN';
+    const createdBy = pullRequest.author;
+    const isApproved = pullRequest.approvals.some(
```
Comment (d4rkr00t) [inline src/Client.js]: Later we can add number of approvals, now it's at least 1, but maybe we want at least 2 :)

File: src/hosts/BitbucketAdapter.js [hunk 10]
```diff
@@ -129,21 +100,43 @@ const BitbucketAdapter = (config: Config) => {
...
+        pullRequestId: pullRequestId,
+        title: data.title,
+        description: data.description,
+        createdOn: new Date(data.created_on),
+        author: data.author.username,
+        state: data.state,
+        approvals: approvals,
+        openTasks: data.task_count
```
Comment (d4rkr00t) [inline src/hosts/BitbucketAdapter.js]: What about GitHub, you want to just put 0 there in the case of GH?

File: src/Client.js (hunk 9 above)
Comment (lukebatchelor) [reply src/Client.js]: Sweet, yup

File: src/hosts/BitbucketAdapter.js (hunk 10 above)
Comment (lukebatchelor) [reply src/hosts/BitbucketAdapter.js]: Yea, that would be nice

File: src/types.js (hunk 7 above)
Comment (lukebatchelor) [reply src/types.js]: Note: this is done now

**Pull request merged**